### ISSUE TITLE
Compatibility with Android/iOS AOT compilation

### ISSUE THREAD

**Issue by breenbob** [dadhi/DryIoc#552]
Prior to the removal of the .WithoutFastExpressionCompiler rule, DryIoc used to work fine in Xamarin.Android, Xamarin.iOS, and I suspect .Net Maui Android/iOS apps also. The introduction of fast expression compilation seems (to me) to be the point at which code that depended on availability of the JIT compiler to run was introduced.

As per the documentation on [limitations](https://learn.microsoft.com/en-gb/xamarin/ios/internals/limitations) of Xamarin.iOS, iOS apps must use AOT compilation and as such do not support aspects of generics and reflection that depend on JIT compilation.

I have run into [this issue](https://github.com/PrismLibrary/Prism.Maui/issues/118) when using the Prism.DryIoc.Maui package:

```
Attempting to JIT compile method 
'FastExpressionCompiler.LightExpression.ExpressionCompiler:TryCompile (FastExpressionCompiler.LightExpression.ExpressionCompiler/ClosureInfo&,
    System.Type,System.Type[],
    System.Type,FastExpressionCompiler.LightExpression.Expression,
    System.Collections.Generic.IReadOnlyList`1<FastExpressionCompiler.LightExpression.ParameterExpression>,bool)' 
while running in aot-only mode. 
See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.
```

The issue itself seems to stem from DryIoc and the FastCompilers packages in their dependency on these unsupported aspects.

This documentation also suggests marking methods that rely on these with one of the following attributes would bypass this problem:
[UnmanagedFunctionPointerAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedfunctionpointerattribute) (preferred, since it is cross-platform and compatible with .NET Standard 1.1+)
[MonoNativeFunctionWrapperAttribute](https://learn.microsoft.com/en-us/dotnet/api/objcruntime.mononativefunctionwrapperattribute)

It states that failing to provide one of these attributes will result in a runtime error exactly like I am seeing in my iOS app above.

As an experiment I have tried [forking DryIoc](https://github.com/breenbob/DryIoc/tree/5.1.0-fast-compiler) from v5.1.0 (version referenced by Prism.DryIoc.Maui) and re-adding the dropped commit for .WithoutFastExpressionCompiler. This was painstaking to say the least! I pushed the packed nuget for this to a private repo and referenced from there. After updating Prism.DryIoc.Maui and my app to use the updated packages, my app _still_ crashed with a similar error to above... but it was a different method this time, nothing to do with the fast compiler stuff:

```
 'Attempting to JIT compile method '(wrapper delegate-invoke) DryIoc.IScope <Module>:invoke_callvirt_IScope_IResolverContext (DryIoc.IResolverContext)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.
```

So I am guessing DryIoc now has other areas in the code also dependency on JIT compilation since this feature was removed that these changes alone do not cover.

Is it possible for support for AOT compilation to be added to DryIoc?

Note, best I can tell this issue does not seem to occur in Prism.Forms, which references DryIoc v4.7.7.

Any help or hints greatly appreciated!

**Comment by dadhi** [comment]
@breenbob Hi, thanks for thorough analysis. 
I am struggling to find the cause before, so hopefully I will find something now with your details.

**Comment by breenbob** [comment]
To be honest I was aware of the limitations around AOT/generics, but I wasn't aware of those attributes until I went searching for the docs to link here. I will give them a go in my fork of 5.1.0 and let you know on here how that works out for me. Thanks for considering this issue, and the great library!

**Comment by breenbob** [comment]
Theoretically, if they work, I am guessing it would mean no need to re-introduce `WithoutFastExpressionCompiler` functionality, but instead updating the FastExpressionCompiler library to use same annotations..

**Comment by dadhi** [comment]
@breenbob Huh, sorry :-( I was for some reason assuming that `WithUseInterpretation()` is used.
Yes, this is the way to solve it.

In the related issues, 
I wanted to improve the situation by recognizing the target is not supporting `System.Reflection.Emit` 
and setting it in the container automatically.
But I did not find the way.

Maybe you will have a suggestion as someone really working on these targets.
It may be some environment check, some reflection probe for missing property, method or class... or preferably more robust modern way?

**Comment by breenbob** [comment]
Yes sorry, I didn't write the package and not really looked at this before.

I'm honestly not sure if there is anyway to detect AOT compilation, either directly or by proxy. I do know there is a csproj setting that gets enabled for this by default in Release build configurations:
`<RunAOTCompilation>true</RunAOTCompilation>`

But doubt there is any way to check that at runtime. Perhaps some msbuild magic...

Whether you can check for unsupported aspects using reflection, I just don't know, but because of AOT compilation coming into play again I would doubt it.

All I can suggest (for Xamarin/.Net Maui at least) is to do something like this:
```cs
#if RELEASE && (ANDROID || IOS || MACCATALYST)
defaultRules = defaultRules.WithUseInterpretation();
#endif
```
But that would not be fool proof, e.g. a custom Release build configuration where the standard Release constant has not been defined. We do this sometimes to have separate configs for CI vs App Store releases...

**Comment by dadhi** [comment]
@breenbob 
This one looks promising:

```cs
#if RELEASE && (ANDROID || IOS || MACCATALYST)
defaultRules = defaultRules.WithUseInterpretation();
#endif
```

Maybe `RELEASE` is not required here, because why care for the compilation in the `DEBUG`. I think from the debugging perspective it even better to have the interpretation, because it is more consistent experience (no compilation jump). 

Then the next question, how to approach it when building the DryIoc library.
- One possibility is to build for the specific target, e.g. `net6.0-ios`, `net6.0-android`, [etc](https://learn.microsoft.com/en-us/dotnet/standard/frameworks).
- The other is to provide some source-generator magic, and have this `#if` in the `partial class Container` implementation on the client side.
The both approaches may have a problem when DryIoc is the transitive dependency of the other library, like the Prism.MAUI.

What are your thoughts on this?
What targets are you using personally, what .net?

**Comment by kyurkchyan** [comment]
For those who may come across this thread. I found a solution with Prism.Maui and DryIoc.

Luckily there's an overload of UserPrism which takes the DryIoc Rules. So, I created rules and passed them to that extension method. IOS/Android apps in Release/AOT mode are not crashing anymore.

```C#
private static Rules DefaultRules => Rules.Default.WithConcreteTypeDynamicRegistrations(reuse:Reuse.Transient)
                                                  .WithUseInterpretation();
                                                  
.....

var builder = MauiApp.CreateBuilder();
            builder
                .UseMauiApp<App>()
                .UsePrism(DefaultRules, Startup.SetupPrismApplication)
....

```

**Comment by dadhi** [comment]
Keeping it open until the docs are updated with the solution.
